internal/api/kubeclient.go

```diff
@@ -0,0 +1,76 @@
+package api
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/akuity/kargo/internal/os"
+)
+
+// TODO: Document this
+func (s *server) getClient(
+	ctx context.Context,
+	namespace string,
+) (client.Client, error) {
+	restCfg, err := getRestConfig(ctx, os.GetEnv("KUBECONFIG", ""), namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "error loading REST config")
+	}
+	return client.New(restCfg, client.Options{Scheme: s.scheme})
+}
+
+// TODO: Document this
+func (s *server) getDynamicClient(
+	ctx context.Context,
+	namespace string,
+) (dynamic.Interface, error) {
+	restCfg, err := getRestConfig(ctx, os.GetEnv("KUBECONFIG", ""), namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, "error loading REST config")
+	}
+	return dynamic.NewForConfig(restCfg)
+}
+
+// TODO: Document this
+//
+// TODO: Test this
+func getRestConfig(
+	ctx context.Context,
+	path string,
+	// We don't use the namespace param yet, but we will
+	namespace string, // nolint: revive, unparam
+) (*rest.Config, error) {
+	var cfg *rest.Config
+	var err error
+	if path == "" {
+		// clientcmd.BuildConfigFromFlags will fall back on in-cluster config if
+		// path is empty, but will issue a warning that we can suppress by checking
+		// for that condition ourselves and calling rest.InClusterConfig() directly.
+		if cfg, err = rest.InClusterConfig(); err != nil {
+			return nil, errors.Wrap(err, "error loading in-cluster REST config")
+		}
+	} else {
+		if cfg, err = clientcmd.BuildConfigFromFlags("", path); err != nil {
+			return nil, errors.Wrapf(err, "error loading REST config from %q", path)
+		}
+	}
+
+	// If, somehow, no credential is found in the context, that's fine. It
+	// shouldn't happen, but in the event that it has, we don't want the client
+	// built from this configuration to use the API server's own service account.
+	//
+	// TODO: This currently only supports client auth. We must add logic here
+	// to map users who have authenticated via other means to a service account.
+	u, _ := userInfoFromContext(ctx)
+	cfg.BearerToken = u.bearerToken
+
+	// This MUST be blanked out because it seems to take precedence over the
+	// cfg.BearerToken field.
+	cfg.BearerTokenFile = ""
+	return cfg, nil
```

Comment:
Can we already support admin accounts here given if it's admin it will directly use the api server token?

Reply:
Yes... this section of code is still WIP.